lib/ssi_processor.rb

```diff
@@ -27,40 +29,43 @@ def process_block(part)
         ""
       end
     end
-    
+
     def process_include(part, blocks)
       part.gsub(/<!--\s?#\s?include\s+(?:virtual|file)="([^"]+)"(?:\s+stub="(\w+)")?\s+-->/) do
         location, stub = $1, $2
         _info "processing include directive with location=#{location}"
         status, _, body = fetch location
         if stub && (status != 200 || body.nil? || body == "")
-          blocks[stub] 
+          blocks[stub]
         else
           body.force_encoding(part.encoding)
         end
       end
     end
-    
+
     def fetch(location)
-      locations.select{|k,v| k.is_a?(String)}.each do |pattern, host|
-        return _get("#{host}#{location}") if location == pattern
-      end
-      locations.select{|k,v| k.is_a?(Regexp)}.each do |pattern, host|
-        return _get("#{host}#{location}") if location =~ pattern
+      options[:locations].each do |pattern, host|
+        next unless pattern === location
+        target = if host.is_a?(Proc)
+          host.call(location)
+        else
+          "#{host}#{location}"
+        end
+        return _get(target)
       end
```

Comment:
I've removed duplication, but exact matches are not prioritized now. I hope users can define priority when passing `:locations` option.